# HG changeset patch
# User Rik <rik@octave.org>
# Date 1529443396 25200
#      Tue Jun 19 14:23:16 2018 -0700
# Node ID dffd9f6ee85ce073499caf7d44833e619f3e4760
# Parent  38a07d930ccdc92f4a640578692b2b4ea177163b
fsolve.m: exit with info=-2 when singularity reached (bug #53991).

* fsolve.m: Check whether the Jacobian is essentially zero, which means the
next update would be the same as the existing solution, and exit with info=-2
if found.

diff --git a/scripts/optimization/fsolve.m b/scripts/optimization/fsolve.m
--- a/scripts/optimization/fsolve.m
+++ b/scripts/optimization/fsolve.m
@@ -299,20 +299,28 @@ function [x, fvec, info, output, fjac] =
     nsuc = 0;
     decfac = 0.5;
 
     ## Inner loop.
     while (niter <= maxiter && nfev < maxfev && ! info)
 
       ## Get trust-region model (dogleg) minimizer.
       if (useqr)
+        if (norm (r, 1) < macheps * rows (r))
+          info = -2;
+          break;
+        endif
         qtf = q'*fvec;
         s = - __dogleg__ (r, qtf, dg, delta);
         w = qtf + r * s;
       else
+        if (norm (fjac, 1) < macheps * rows (fjac))
+          info = -2;
+          break;
+        endif
         s = - __dogleg__ (fjac, fvec, dg, delta);
         w = fvec + fjac * s;
       endif
 
       sn = norm (dg .* s);
       if (niter == 1)
         delta = min (delta, sn);
       endif
